### PR TITLE
Move dynamic FPU synchronization inside preprocessor defines

### PR DESCRIPTION
**Pull request by kcgen**
This is a follow-up adjustment mentioned by @finalpatch; thanks!
 
This PR also:
- Makes non-functional adjustments to the previously-touched FPU additions (uses more const and auto arguments, replaces C casts with C++ casts, uses default initializes, and uses C++ struct definition style ("struct Type {... }", just like classes).
- Fixes a couple dangling memset/memcpy warnings by using the unaligned accessor functions.
- Uses the Get and Set verbs in the PReg function names (to make the match the others).
- The last two commits are functional changes (deliberately independent and the last in this set) to be easy bisection points if either cause problems down the road.

Hoping you can test this with your Turbo C debugger @finalpatch (and anything else you can throw at it like large and tiny values / overflow / div-by-zero / NaN / etc).

Suggest reviewing commit-by-commit.

DOS Navigator is still OK, as are UBSAN builds.

### CODE REVIEW

File: src/cpu/core_dyn_x86.cpp [hunk 1]
```diff
@@ -263,25 +275,36 @@ static void copy_normal_fpu_to_dh()
...
+{
+	return CPU_Core_Normal_Run();
+}
+#	endif
+
 Bits CPU_Core_Dyn_X86_Run() noexcept
 {
 	ZoneScoped;
-	// helper class to auto-save DH_FPU state on function exit
-	class auto_dh_fpu {
```
Comment (finalpatch) [inline src/cpu/core_dyn_x86.cpp:273]: No you can't delete this either.  This makes sure whenever we leave CPU_Core_Dyn_X86_Run(),  we keep a snapshot of the dynrec's FPU state so later we can resume.   I would also keep the destructor pattern because this function has many exits, remember to manually saving states at each place is going to be a maintenance burden.

File: src/cpu/core_dyn_x86.cpp [hunk 2]
```diff
@@ -310,30 +330,7 @@ Bits CPU_Core_Dyn_X86_Run() noexcept
...
-			// block, then we need to save the host FPU state in
-			// dyn_dh_fpu.state and free up the host FPU for the
-			// normal core to use.
-			if (dyn_dh_fpu.state_used) {
-				gen_dh_fpu_save();
-			}
-
-			// Copy the saved state to the normal core's emulated
-			// FPU so that it can resume from where the dynamic core
-			// left. We can't skip this step even when
```
Comment (finalpatch) [inline src/cpu/core_dyn_x86.cpp:323]: Here I explained why state_used cannot be used to skip this copy.  We should keep this comment in the code (maybe make it even more obvious) so that people do not try to mistakenly optimize this code.

File: src/fpu/fpu.cpp [hunk 3]
```diff
@@ -27,39 +27,51 @@
...
+{
+	uint16_t tag = 0;
+	for (Bitu i = 0; i < 8; i++) {
+		tag |= ((fpu.tags[i] & 3) << (2 * i));
+	}
 	return tag;
 }
 
-void FPU_LoadPRegs(const uint8_t* buffer)
+void FPU_SetPRegsFrom(const uint8_t dyn_regs[8][10])
```
Comment (finalpatch) [inline src/fpu/fpu.cpp:46]: I think the original method of copying from a typeless buffer is okay.  This is essentially copying from a buffer saved by the FSAVE machine instruction.  There is no point to attempt to give it some structure.  The compiler can optimize fixed size memcpy very effectively (https://godbolt.org/z/Pv3fch5fM)

File: src/fpu/fpu.cpp [hunk 4]
```diff
@@ -27,39 +27,51 @@
...
+		const auto dyn_reg_addr = dyn_regs[i];
+
+		norm_p_reg.m1 = read_unaligned_uint32_at(dyn_reg_addr, 0);
+		norm_p_reg.m2 = read_unaligned_uint32_at(dyn_reg_addr, 1);
+		norm_p_reg.m3 = read_unaligned_uint16_at(dyn_reg_addr, 4);
 	}
 }
 
-void FPU_SavePRegs(uint8_t* buffer)
+void FPU_GetPRegsTo(uint8_t dyn_regs[8][10])
```
Comment (finalpatch) [inline src/fpu/fpu.cpp:58]: Same, I recommend sticking to the original code both for simplicity and performance.

**Comment by finalpatch** [comment]
Honestly I would recommend just do these 2 changes and minimize the change to the original code.  This way the risk is kept to the minimum (this code is very delicate and I would very cautious to make any change that I'm not 100% confident with) and it's easy for other dosbox forks to pick up this fix rather than having to deal with a large diff without lots of unrelated changes.

**Comment by Grounded0** [comment]
I'm for minimising risk since this is extremely delicate piece of code.

File: include/fpu.h [hunk 5]
```diff
@@ -45,110 +45,111 @@ void FPU_ESC6_EA(Bitu func,PhysPt ea);
...
 
-
-uint16_t FPU_GetTag(void);
+uint16_t FPU_GetTag();
 void FPU_FLDCW(PhysPt addr);
 
-static inline void FPU_SetTag(uint16_t tag){
-	for(Bitu i=0;i<8;i++)
-		fpu.tags[i] = static_cast<FPU_Tag>((tag >>(2*i))&3);
+static inline void FPU_SetTag(const uint32_t tag)
```
[the remark this answers is not in the document]
Comment (kcgen) [reply include/fpu.h]: The dh's FPU tag is 32bit:

``` c++
struct dyn_dh_fpu {
	// ...
	struct {
		uint32_t cw;
		uint32_t sw;
		uint32_t tag;
		uint8_t st_reg[8][10];
	} state;
	// ...
```

and that's what's being passed in:

```c++
FPU_SetTag(dyn_dh_fpu.state.tag);
```

This causes an implicit cast across the API, and will wrap-around of the value if there happens to be data beyond the 16-bit threshold.

Another option is to fix the size to portion used, and then expect the caller to reduce the content inbound: so I've done that now:

``` c++
FPU_SetTag(static_cast<uint16_t>(dyn_dh_fpu.state.tag & 0xffff));
```

Which let's us maintain the `uint16_t` in bound.

Thanks for flagging it :+1:

File: include/fpu.h [hunk 6]
```diff
@@ -45,110 +45,111 @@ void FPU_ESC6_EA(Bitu func,PhysPt ea);
...
+	}
 }
 
 static inline uint16_t FPU_GetCW()
 {
 	return fpu.cw;
 }
 
-static inline void FPU_SetCW(Bitu word)
+static inline void FPU_SetCW(const uint32_t dword)
```
[the remark this answers is not in the document]
Comment (kcgen) [reply include/fpu.h]: Same as above: the dh's FPU state `cw` is fixed to 32bit:

``` c++
struct dyn_dh_fpu {
	// ...
	struct {
		uint32_t cw;
		uint32_t sw;
		uint32_t tag;
		uint8_t st_reg[8][10];
	} state;
	// ...
```

On 64-bit platforms, it's being unnecessarily cast up to 64-bit.

When DOSBox started, 32-bit was the native pointer size (on Windows 2000 or maybe XP), and `Bitu` was used a bit too liberally, even for pure numeric value-types (and registers, like this here), which don't involve carrying around host-pointers as values.

Whenever we touch existing code, we try to de-Bitu it as well (when it makes sense.. there's still a lot left, and many are perfectly fine). This is actually one of the areas QBix suggested we consider fixing, too. 

Originally, we had roughly 2000 Bitu-related warnings, but over time, we've got it down, now in the 800s.




This is what folks building on 64-bit MSVC currently see. Welcome to the firehose :sweat_smile:

[the remark this answers is not in the document]
Comment (kcgen) [reply include/fpu.h]: I've gone down from Bitu to uint16_t now, and am chopping off the excess on the caller-side:

```c++
FPU_SetCW(static_cast<uint16_t>(dyn_dh_fpu.state.cw & 0xffff));
```

File: include/fpu.h [hunk 7]
```diff
@@ -45,110 +45,111 @@ void FPU_ESC6_EA(Bitu func,PhysPt ea);
...
+	fpu.round       = static_cast<FPU_Round>((word >> 10) & 3);
 }
 
 static inline uint16_t FPU_GetSW()
 {
 	return fpu.sw;
 }
 
-static inline void FPU_SetSW(Bitu word)
+static inline void FPU_SetSW(const uint32_t dword)
```
[the remark this answers is not in the document]
Comment (kcgen) [reply include/fpu.h]: Same reason as above: to match inbound type-size, and let the received truncate.

Same resolution: moved to caller-size type limiting:

```c++
FPU_SetSW(static_cast<uint16_t>(dyn_dh_fpu.state.sw & 0xffff));
```

File: src/fpu/fpu.cpp (hunk 3 above)
Comment (kcgen) [reply src/fpu/fpu.cpp:46]: > I think the original method of copying from a typeless buffer is okay. 

No argument; it gets the job done.

Unfortunately only partially writing structs and class data (via bulk memcpy and memset) often causes compilers to caution that we might be leaving stuff behind. 





It's a good warning because when someone changes a struct definition, the impact can ripple to all the memcpy/memset's that are using hard-coded quantities.

The classic solution is using sizeof on the struct to set memcpy's quantity to keep it in lock-step with the definition. 

And no doubt, you would have done that here too, but I realize we're only after a deliberate subset of the struct members.

Under the hood, the `unaligned_` accessors also use memcpy to access the bytes and cast up the result to bonafide types: and like you, confirmed that those memcpy's get translated to essentially the same code as the memory type-casters (so zero performance impact).

In this case: I realize these aren't available on other DOSBoxes: so I suggest the patches still use the memcpy (and those projects can assess those warnings).

Of course, they are also welcome to grab Staging's unaligned accessors. They are tiny and extensively tested.

```text
include/mem_unaligned.h:static inline uint16_t read_unaligned_uint16(const uint8_t *arr) noexcept
include/mem_unaligned.h:static inline uint32_t read_unaligned_uint32(const uint8_t *arr) noexcept
include/mem_unaligned.h:static inline uint64_t read_unaligned_uint64(const uint8_t *arr) noexcept
include/mem_unaligned.h:static inline size_t read_unaligned_size_t(const uint8_t *arr) noexcept
include/mem_unaligned.h:static inline uint16_t read_unaligned_uint16_at(const uint8_t *arr,
include/mem_unaligned.h:static inline uint32_t read_unaligned_uint32_at(const uint8_t *arr,
include/mem_unaligned.h:static inline uint64_t read_unaligned_uint64_at(const uint8_t *arr,
include/mem_unaligned.h:static inline size_t read_unaligned_size_t_at(const uint8_t *arr,
include/mem_unaligned.h:static inline void write_unaligned_uint16(uint8_t *arr, const uint16_t val) noexcept
include/mem_unaligned.h:static inline void write_unaligned_uint32(uint8_t *arr, const uint32_t val) noexcept
include/mem_unaligned.h:static inline void write_unaligned_uint64(uint8_t *arr, const uint64_t val) noexcept
include/mem_unaligned.h:static inline void write_unaligned_uint16_at(uint8_t *arr,
include/mem_unaligned.h:static inline void write_unaligned_uint32_at(uint8_t *arr,
include/mem_unaligned.h:static inline void write_unaligned_uint64_at(uint8_t *arr,
include/mem_unaligned.h:static inline void add_to_unaligned_uint16(uint8_t *arr, const uint16_t val) noexcept
include/mem_unaligned.h:static inline void add_to_unaligned_uint32(uint8_t *arr, const uint32_t val) noexcept
include/mem_unaligned.h:static inline void add_to_unaligned_uint64(uint8_t *arr, const uint64_t val) noexcept
include/mem_unaligned.h:static inline void inc_unaligned_uint16(uint8_t *arr) noexcept
include/mem_unaligned.h:static inline void inc_unaligned_uint32(uint8_t *arr) noexcept
include/mem_unaligned.h:static inline void inc_unaligned_uint64(uint8_t *arr) noexcept
```

File: src/fpu/fpu.cpp (hunk 4 above)
Comment (kcgen) [reply src/fpu/fpu.cpp:58]: (see above for same reason / recommendation).

I think a safe approach will be to use extract a patch from DOSBox-X, as that should be compatible w/ SVN too.

File: src/cpu/core_dyn_x86.cpp [hunk 8]
```diff
@@ -263,25 +275,36 @@ static void copy_normal_fpu_to_dh()
...
+	FPU_GetPRegsTo(dyn_dh_fpu.state.st_reg);
 }
 
+static Bits sync_fpu_and_run_normal_core() noexcept
+{
+	if (dyn_dh_fpu.state_used) {
+		// If the host's FPU was used by a recompiled code block, we
+		// copy its state into dyn_dh_fpu.state and from there into the
+		// normal core's emulated FPU
+		copy_dh_fpu_to_normal();
```
[the remark this answers is not in the document]
Comment (kcgen) [reply src/cpu/core_dyn_x86.cpp]: Thanks, @finalpatch.

I realize my original testing in the PR was when the dh-to-normal copy code was only wrapped around that single call to `CPU_Core_Normal_Run()` (before it was added around all the other calls, too).

So I went back to the benchmarks and added more logging, and indeed, these FPU copies can now happen in excess of 100 times per DOSBox tick (per millisecond), as a function of cycle count.

For example:



That's all within the same tick (note, this is with skipping.. more below :slightly_smiling_face: !)

File: src/cpu/core_dyn_x86.cpp (hunk 2 above)
Comment (kcgen) [reply src/cpu/core_dyn_x86.cpp:323]: > Here I explained why state_used cannot be used to skip this copy. 

Indeed - and with logging, I've also come to the same conclusion, and have left in the comment. Thanks.

After some digging, we can add another dirty-tracker used to  indicate if dh needs to be copied to the emulated normal FPU.

Whenever the dynamic core uses the host FPU, it sets the `state_used` to `true`.  The only way this gets flipped back to `false` is when `gen_dh_fpu_save()` is called (existing code):

``` c++
#if defined(X86_DYNFPU_DH_ENABLED)
static void gen_dh_fpu_save(void)
#if defined (_MSC_VER)
{
	__asm {
	__asm	fnsave	dyn_dh_fpu.state
	__asm	fldcw	dyn_dh_fpu.host_cw
	}
	dyn_dh_fpu.state_used=false;
	dyn_dh_fpu.state.cw|=0x3f;
}
#else
{
	__asm__ volatile (
		"fnsave		%0			\n"
		"fldcw		%1			\n"
		:	"=m" (dyn_dh_fpu.state)
		:	"m" (dyn_dh_fpu.host_cw)
		:	"memory"
	);
	dyn_dh_fpu.state_used=false;
	dyn_dh_fpu.state.cw|=0x3f;
}
#endif
#endif

```

The PR adds `maybe_copy_host_fpu_to_dh()` that sets this new dirty tracker:

``` c++
void static void maybe_copy_host_fpu_to_dh()
{
	if (dyn_dh_fpu.state_used) {
		gen_dh_fpu_save();
		dyn_dh_fpu.state_copied_to_normal = false;
	}
}
```

Which now called both on the scope closures (just like the original class did, but named to fall in line with the other copy functions):

``` c++
struct HostFpuToDhCopier {
	~HostFpuToDhCopier()
	{
		maybe_copy_host_fpu_to_dh();
	}
};
```

as well as on entry to the normal core:

``` c++
static Bits copy_fpus_and_run_normal_core() noexcept
{
	// If host FPU has been used by the recompiled code
	// block, then we need to save the host FPU state in
	// dyn_dh_fpu.state and free up the host FPU for the
	// normal core to use.

 	// if (dyn_dh_fpu.state_used) {
	//	LOG_WARNING("FPU: copying host to dh and using normal core (normal FPU now invalid)");
	// }
	maybe_copy_host_fpu_to_dh();

	// If the dynamic core's FPU state was changed and hasn't yet been
	// copied to the normal core, then copy the dyn_dh_fpu.state to the
	// normal core's emulated FPU so that it can resume from where the
	// dynamic core left.
	maybe_copy_dh_fpu_to_normal();

	// LOG_MSG("FPU: Entering normal core (most recent normal FPU state should be valid)");
	auto ret = CPU_Core_Normal_Run();
```

And as you've no doubt guessed:

``` c++
static void maybe_copy_dh_fpu_to_normal()
{
	if (dyn_dh_fpu.state_copied_to_normal) {
		// LOG_MSG("FPU: skipping dh to normal because dh hasn't changed (normal FPU still valid)");
		return;
	}

	// Pre-requisite: host FPU should be copied to the dh FPU
	assert(!dyn_dh_fpu.state_used);

	FPU_SetTag(static_cast<uint16_t>(dyn_dh_fpu.state.tag & 0xffff));
	FPU_SetCW(static_cast<uint16_t>(dyn_dh_fpu.state.cw & 0xffff));
	FPU_SetSW(static_cast<uint16_t>(dyn_dh_fpu.state.sw & 0xffff));
	TOP = FPU_GET_TOP();
	FPU_SetPRegsFrom(dyn_dh_fpu.state.st_reg);

	dyn_dh_fpu.state_copied_to_normal = true;
	// LOG_MSG("FPU: copying dh to normal because dh has changed (normal FPU now valid)");
}
```

But what about the outbound normal-to-dh copy that still always happens, doesn't that mean the dh side is dirty again?

There are two pathways:

If the normal core /did/ changed its FPU state, then that state is written back out to the DH side - so the DH side is now fully in-sync with the normal side.

If the normal core /didn't/ change its FPU state, then that state is written back out anyway (a wasted step).

On the next round(s) in the dynamic core, so long as the FPU state isn't touched, then the normal side is still in sync and we can keep skipping.

Here are some log results:

File: src/cpu/core_dyn_x86.cpp [hunk 9]
```diff
@@ -248,13 +257,34 @@ static void dyn_restoreregister(DynReg * src_reg, DynReg * dst_reg) {
...
-	FPU_SetSW(dyn_dh_fpu.state.sw);
+	if (dyn_dh_fpu.state_used) {
+		gen_dh_fpu_save();
+		dyn_dh_fpu.state_copied_to_normal = false;
+	}
+}
+
+static void maybe_copy_dh_fpu_to_normal()
+{
+	if (dyn_dh_fpu.state_copied_to_normal) {
```
Comment (finalpatch) [inline src/cpu/core_dyn_x86.cpp:272]: I have a PR in the X repo that utilises similar idea.  But I perform the copies at the 2 choke points:
* entering normal core
* entering dynamic core

File: src/cpu/core_dyn_x86.cpp [hunk 10]
```diff
@@ -248,13 +257,34 @@ static void dyn_restoreregister(DynReg * src_reg, DynReg * dst_reg) {
...
+	FPU_SetSW(static_cast<uint16_t>(dyn_dh_fpu.state.sw & 0xffff));
 	TOP = FPU_GET_TOP();
-	FPU_LoadPRegs(&dyn_dh_fpu.state.st_reg[0][0]);
+	FPU_SetPRegsFrom(dyn_dh_fpu.state.st_reg);
+
+	dyn_dh_fpu.state_copied_to_normal = true;
+	// LOG_MSG("FPU: copying dh to normal because dh has changed (normal FPU now valid)");
 }
 
 static void copy_normal_fpu_to_dh()
```
Comment (finalpatch) [inline src/cpu/core_dyn_x86.cpp:290]: It's not safe to unconditionally copy normal fpu back.

File: src/cpu/core_dyn_x86.cpp (hunk 9 above)
Comment (kcgen) [reply src/cpu/core_dyn_x86.cpp:272]: Thanks for the heads up, @finalpatch 👍

File: src/cpu/core_dyn_x86.cpp (hunk 10 above)
Comment (finalpatch) [reply src/cpu/core_dyn_x86.cpp:290]: We can only copy normal fpu back when we previously copied it in.  Otherwise we risk overwriting dynamic fpu with unknown state.

File: include/fpu.h [hunk 11]
```diff
@@ -45,105 +45,105 @@ void FPU_ESC6_EA(Bitu func,PhysPt ea);
...
+	uint16_t cw_mask_all = 0;
+	uint16_t sw          = 0;
+	uint32_t top         = 0;
+	FPU_Round round      = {};
+};
 
+constexpr auto L2E = 1.4426950408889634;
+constexpr auto L2T = 3.3219280948873623;
+constexpr auto LN2 = 0.69314718055994531;
+constexpr auto LG2 = 0.3010299956639812;
```
Comment (johnnovak) [inline include/fpu.h:99]: While you're at it, can get sane names for these? Took me almost 10 mins to reverse engineer them 😅 

```
constexpr auto Log2_E = 1.4426950408889634;
constexpr auto Ln10DivLn2 = 3.3219280948873623;
constexpr auto Ln2 = 0.69314718055994531;
constexpr auto Log10_2 = 0.3010299956639812;
```

File: include/fpu.h [hunk 12]
```diff
@@ -45,105 +45,105 @@ void FPU_ESC6_EA(Bitu func,PhysPt ea);
...
+{
+	return static_cast<uint8_t>((fpu.sw & 0x3800) >> 11);
 }
 
-static inline void FPU_SET_TOP(Bitu val){
+static inline void FPU_SET_TOP(const uint32_t val)
+{
 	fpu.sw &= ~0x3800;
-	fpu.sw |= (val&7)<<11;
+	fpu.sw |= static_cast<uint16_t>((val & 7) << 11);
```
Comment (johnnovak) [inline include/fpu.h:146]: I don't think this helps with clarity. Why not extract `0x3800` into a const instead and reference in both the setter and getter?

File: src/cpu/core_dyn_x86.cpp [hunk 13]
```diff
@@ -275,21 +287,55 @@ static void copy_normal_fpu_to_dh()
...
+
+	auto ret = CPU_Core_Normal_Run();
+
+	// On leaving the normal core, copy emulated FPU's state
+	// to the dynamic core's saved FPU state. There is no
+	// way to tell if the emulated FPU was used in this run
+	// so we can't skip this.
+	copy_normal_fpu_to_dh();
+
+	return ret;
```
Comment (johnnovak) [inline src/cpu/core_dyn_x86.cpp:326]: Nice explanatory comments 💯

File: src/cpu/core_dyn_x86.cpp [hunk 14]
```diff
@@ -263,19 +263,28 @@ static void maybe_copy_host_fpu_to_dh()
...
+		dyn_dh_fpu.state_copied_to_normal = false;
 	}
 }
 
-static void copy_dh_fpu_to_normal()
+static void maybe_copy_dh_fpu_to_normal()
 {
-	// The host's FPU should already be copied to the dh FPU state at this point
+	if (dyn_dh_fpu.state_copied_to_normal) {
+		LOG_MSG("FPU: skipping dh to normal because dh hasn't changed (normal FPU still valid)");
```
Comment (johnnovak) [inline src/cpu/core_dyn_x86.cpp]: All these should be debug or trace level, no? Regular users shouldn't see things like these in the logs. Also, what if these get called frequently? We don't want to spam the logs and slow things down.

Same comment for all log statements in this file.

File: src/cpu/core_dyn_x86.cpp [hunk 15]
```diff
@@ -328,9 +329,9 @@ static Bits copy_fpus_and_run_normal_core() noexcept
 struct HostFpuToDhCopier {
 	~HostFpuToDhCopier()
 	{
-		if (dyn_dh_fpu.state_used) {
-			LOG_WARNING("FPU: copying host to dh but not using normal core (normal FPU now invalid)");
-		}
+		// if (dyn_dh_fpu.state_used) {
+		//	LOG_WARNING("FPU: copying host to dh but not using normal core (normal FPU now invalid)");
+		// }
```
Comment (johnnovak) [inline src/cpu/core_dyn_x86.cpp:334]: Okay, you commented them out here... I'm reviewing commit by commit 😄

**Comment by johnnovak** [comment]
> Honestly I would recommend just do these 2 changes I suggested earlier and minimize the change to the original code. This way the risk is kept to the minimum (this code is very delicate and I would be very cautious to make any change that I'm not 100% confident with) and it's easy for other dosbox forks to pick up this fix rather than having to deal with a large diff with lots of unrelated changes.

> I'm for minimising the risk since this is an extremely delicate part of the codebase.

I agree with both of you on the need to test this very thoroughly. There must be some benchmark apps out there that test the FPU for correctness. I'd also be uncomfortable merging this just based on a "visual" code review — there's so many things that can go wrong.

Having said that, we shouldn't be afraid of changing the code; that's a bad situation to be in. So instead of never touching these parts of the codebase, the question we should be asking is how can be regressions test this to gain 100% confidence that it still works as expected? Like I said, someone must have written some FPU test benchmarks in the DOS days...

**Comment by finalpatch** [comment]
> I agree with both of you on the need to test this very thoroughly. There must be some benchmark apps out there that test the FPU for correctness. I'd also be uncomfortable merging this just based on a "visual" code review — there's so many things that can go wrong.

What's difficult is that this code deals with the case when FPU is used cross core boundary, which AFAIK happens when code modifies itself or when debugger/OS doing system level operations.  A benchmark suite normally stays in the dynamic core happy path and won't see anything wrong.  That's why DOSBOX has been able to run games for so many years without people noticing this.

**Comment by finalpatch** [comment]
I think in order to properly test this, we need to create a guest program in assembly code that deliberately exercises this core transition logic.

**Comment by johnnovak** [comment]
> I think in order to properly test this, we need to create a guest program in assembly code that deliberately exercises this core transition logic.

Ok fair point.

File: include/fpu.h (hunk 11 above)
Comment (kcgen) [reply include/fpu.h]: Added! `cmath` had 'em too :-) 

Switched to the built-ins as `constexpr`s, for the two that involved dividing log 10 / 2 and ln 10 / 2.

File: src/cpu/core_dyn_x86.cpp (hunk 13 above)
Comment (kcgen) [reply src/cpu/core_dyn_x86.cpp]: tried to summaries it more concisely in the latest push (and hopefully the code is more self-documenting too).

File: src/cpu/core_dyn_x86.cpp (hunk 14 above)
Comment (kcgen) [reply src/cpu/core_dyn_x86.cpp]: (drop them in the latest PR)

**Comment by kcgen** [comment]
We've got two bisectable points now:
 - the prior PR, where the normal and dh FPU's are synchronized in every call (no matter what)
 - This PR where they're only syn'd if the opposite core was last used

That's really all we need to track down strange bugs.

The other good thing is that somehow all of eXoDOS and eXoWin3.1 are running fine with this synchronization not even done at all! Granted.. those are all games/edutainment, but still, that's about 15,000 executables.

With @finalpatch's Turbo Debugger tests showing proper operation, hopefully this improvement fixes more application corner cases.

Oh: and we're aware of DOSBox being used for civil engineering work, like bridge and elevated load bearing calculations.   Likewise, there are German manufacturers using DOSBox to run software that drives their fabrication equipment.

So hopefully the numbers generated are still OK! :rocket:

**Comment by finalpatch** [comment]
I'll try to put together a test executable that deterministically exercises this logic. Give me some time.

**Comment by Grounded0** [comment]
> Oh: and we're aware of DOSBox being used for civil engineering work, like bridge and elevated load bearing calculations. Likewise, there are German manufacturers using DOSBox to run software that drives their fabrication equipment.

We're currently offering them 64-bit resolution instead of an actual IEEE 754 80-bit extended precision format so it should be all good. 😁

But yeah, many people are depending that we get the math correct that probably shouldn't which is yet another factor in mitigating risk. 

Maybe we should deploy the safer option per @finalpatch first and this extended option could be deployed later once we're satisfied everything still works with the more limited scope first and a test program for x87 have been written to validate this bit of code.

**Comment by rderooy** [comment]
You can try Intel's MCP Diag.
https://archive.org/details/msdos_MCPDIAG_shareware

But be aware that it will likely fail by default, as it did originally on DOSBox-X.
https://github.com/joncampbell123/dosbox-x/pull/3415

**Comment by finalpatch** [comment]
Okay,  I managed to create a program that can trigger the bug
https://gist.github.com/finalpatch/91ae4449875c3cf217327bfdcbeab0c5

**Comment by finalpatch** [comment]
This branch in its current state produces "ok" from my test program.

File: src/fpu/fpu.cpp (hunk 3 above)
Comment (finalpatch) [reply src/fpu/fpu.cpp:46]: That is  one `memcpy` vs  3 `memcpy` (and maybe some extra assignments)

**Comment by finalpatch** [comment]
The core switch logic as far as I can see, is correct (I still believe 1 memcpy is better than 3 in this performance sensitive code)

 However there are way too many unrelated changes in this PR that I don't have the bandwidth/motivation to review. I'll leave it to the other folks to approve the whole PR,  the part in in core_dyn_x86.cpp looks good to me.

**Comment by Grounded0** [comment]
core_dyn_x86.cpp looks good to me.

**Comment by johnnovak** [comment]
Can't really do a useful review of this myself, but the approach @kcgen mentioned with the easy bisecting, plus the asm test code by @finalpatch has probably raised our confidence levels enough to get this merged?

File: tests/files/fpu_test.asm [hunk 16]
```diff
@@ -0,0 +1,89 @@
+; Authored by finalpatch (GitHub)
+;   Source: https://gist.github.com/finalpatch/91ae4449875c3cf217327bfdcbeab0c5
+;   Tests: https://github.com/dosbox-staging/dosbox-staging/issues/2247
+;
+; tasm fpu
+; tlink fpu
```
Comment (finalpatch) [inline tests/files/fpu_test.asm:6]: this is the assembler/linker instruction.  since you changed filename to fpu_test,  you need to update this as well.

**Comment by kcgen** [comment]
> This branch in its current state produces "ok" from my test program.

Thanks. Do you have the compiled .EXE handy? (I tried w/ `fasm` but it complains about not entering 32-bit mode).  It would be nice to add this to our unit tests that run on the Apple M1, x86-64, ARM v7 and v8, as well as on Windows w/ MSVC. 

(To unit test it, I would launch it with the compiled dosbox-staging executable, redirect to a text file, and check for OK).

> The core switch logic as far as I can see, is correct

Thanks for the review.  

>  (I still believe 1 memcpy is better than 3 in this performance sensitive code)

Yes, I explained in the resolved comments that the type-specific accessors and assigning directly to the struct members is still more self-documenting and explicit versus the bulk memcpy.

But to wrap this up, I dropped the two routines in quick bench:

Here's the assembly at `-O2` (ARM middle-pane, x86-64 right-pane):



At `-O3`, the memcpy approach stays a loop while the approach used in this PR goes branchless (eliminating the loop):



https://godbolt.org/z/seGvx8e5M

The loopless optimized version makes it very hard to benchmark at `-O3` in Quick Bench. Here's Clang 15 and GCC 12.x, at `-O3`:



I tried flagging variables as do-not-optimize, accumulating intermediate p_reg m1/m2/m3, randomizing the dyn_reg array, etc, to no avail.

However, Clang 14 atleast is measuring /something/ :laughing:



https://quick-bench.com/q/LkhSkKnQLmvE_amiu7ixUd5j-kQ

All that to say - we're deep in the noise (a couple instructions either way, and for something that's only called during transitions from core types).  These explorations are always fun though!

File: tests/files/fpu_test.asm (hunk 16 above)
Comment (kcgen) [reply tests/files/fpu_test.asm:6]: > this is the assembler/linker instruction. 

Yes, I just don't have a tasm or tlink (but I do have `fasm`).
If you don't have the binary handy, I guess I'll track these down later.

>  since you changed filename to fpu_test, you need to update this as well

Will do.

Comment (kcgen) [reply tests/files/fpu_test.asm]: In bound.

**Comment by kcgen** [comment]
> Can't really do a useful review of this myself, but the approach @kcgen mentioned with the easy bisecting, plus the asm test code by @finalpatch has probably raised our confidence levels enough to get this merged?

Yes - I think we're set to merge, thanks to @finalpatch's test results and everything else that has been tested (all running fine).

Thanks again for this, @finalpatch - and very interested in any more additions you want to help with! :heart:

**Comment by finalpatch** [comment]
test program binary and source
https://1drv.ms/u/s!As8r3moTN9P9gZcnPBF0cUA9TBkjwg?e=mDaX4w

**Comment by kcgen** [comment]
Nice! thanks for these, @finalpatch.

Anyone know of a CLI test framework (or system)?

This FPU tester will be our first (of probably many) executables we run using the compiled DOSBox Staging and inspect the results.

This is essentially black-box executable to go alongside our current white-box (GoogleTest and GoogleMock) internal test framework.
